// react-avatar/lib/components/image.js
"use strict";

var _Reflect$construct = require("core-js-pure/stable/reflect/construct.js");

var _Object$defineProperty = require("core-js-pure/stable/object/define-property.js");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../utils");

var _wrapper = _interopRequireDefault(require("./wrapper"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = _Reflect$construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !_Reflect$construct) return false; if (_Reflect$construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var AvatarImage = /*#__PURE__*/function (_React$PureComponent) {
  (0, _inherits2.default)(AvatarImage, _React$PureComponent);

  var _super = _createSuper(AvatarImage);

  function AvatarImage() {
    (0, _classCallCheck2.default)(this, AvatarImage);
    return _super.apply(this, arguments);
  }

  (0, _createClass2.default)(AvatarImage, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          round = _this$props.round,
          unstyled = _this$props.unstyled,
          alt = _this$props.alt,
          title = _this$props.title,
          name = _this$props.name,
          value = _this$props.value,
          avatar = _this$props.avatar;
      var size = (0, _utils.parseSize)(this.props.size);
      var imageStyle = unstyled ? null : {
        maxWidth: '100%',
        width: size.str,
        height: size.str,
        borderRadius: (0, _utils.calculateBorderRadius)(round)
      };
      return /*#__PURE__*/_react.default.createElement(_wrapper.default, this.props, /*#__PURE__*/_react.default.createElement("img", {
        className: className + ' sb-avatar__image',
        width: size.str,
        height: size.str,
        style: imageStyle,
        src: avatar.src,
        alt: (0, _utils.getNullableText)(alt, name || value),
        title: (0, _utils.getNullableText)(title, name || value),
        onError: avatar.onRenderFailed
      }));
    }
  }]);
  return AvatarImage;
}(_react.default.PureComponent);

exports.default = AvatarImage;
(0, _defineProperty2.default)(AvatarImage, "propTypes", {
  alt: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.bool]),
  title: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.bool]),
  name: _propTypes.default.string,
  value: _propTypes.default.string,
  avatar: _propTypes.default.object,
  className: _propTypes.default.string,
  unstyled: _propTypes.default.bool,
  round: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string, _propTypes.default.number]),
  size: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string])
});
(0, _defineProperty2.default)(AvatarImage, "defaultProps", {
  className: '',
  round: false,
  size: 100,
  unstyled: false
});
